c++/src/ColumnWriter.cc

```diff
@@ -0,0 +1,507 @@
+      } else {
+        const char* notNull = structBatch.notNull.data() + offset;
+        for (uint64_t i = 0; i < numValues; ++i) {
+          if (notNull[i]) {
+            colIndexStatistics->increase(1);
+          } else if (!hasNull) {
+            hasNull = true;
+          }
+        }
+      }
+      colIndexStatistics->setHasNull(hasNull);
+    }
+  }
+
+  void StructColumnWriter::flush(std::vector<proto::Stream>& streams) {
+    ColumnWriter::flush(streams);
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->flush(streams);
+    }
+  }
+
+  void StructColumnWriter::writeIndex(
+                      std::vector<proto::Stream> &streams) const {
+    ColumnWriter::writeIndex(streams);
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->writeIndex(streams);
+    }
+  }
+
+  uint64_t StructColumnWriter::getEstimatedSize() const {
+    uint64_t size = ColumnWriter::getEstimatedSize();
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      size += children[i]->getEstimatedSize();
+    }
+    return size;
+  }
+
+  void StructColumnWriter::getColumnEncoding(
+                      std::vector<proto::ColumnEncoding>& encodings) const {
+    proto::ColumnEncoding encoding;
+    encoding.set_kind(proto::ColumnEncoding_Kind_DIRECT);
+    encoding.set_dictionarysize(0);
+    encodings.push_back(encoding);
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->getColumnEncoding(encodings);
+    }
+  }
+
+  void StructColumnWriter::getStripeStatistics(
+    std::vector<proto::ColumnStatistics>& stats) const {
+    ColumnWriter::getStripeStatistics(stats);
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->getStripeStatistics(stats);
+    }
+  }
+
+  void StructColumnWriter::mergeStripeStatsIntoFileStats() {
+    ColumnWriter::mergeStripeStatsIntoFileStats();
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->mergeStripeStatsIntoFileStats();
+    }
+  }
+
+  void StructColumnWriter::getFileStatistics(
+    std::vector<proto::ColumnStatistics>& stats) const {
+    ColumnWriter::getFileStatistics(stats);
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->getFileStatistics(stats);
+    }
+  }
+
+  void StructColumnWriter::mergeRowGroupStatsIntoStripeStats()  {
+    ColumnWriter::mergeRowGroupStatsIntoStripeStats();
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->mergeRowGroupStatsIntoStripeStats();
+    }
+  }
+
+  void StructColumnWriter::createRowIndexEntry() {
+    ColumnWriter::createRowIndexEntry();
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->createRowIndexEntry();
+    }
+  }
+
+  void StructColumnWriter::resetIndex() {
+    ColumnWriter::resetIndex();
+
+    for (uint32_t i = 0; i < children.size(); ++i) {
+      children[i]->resetIndex();
+    }
+  }
+
+  class IntegerColumnWriter : public ColumnWriter {
```

Comment:
Not necessarily. There are just no much common logics between different types. We can re-examine this in next PR which includes other type writers.